src/pip/_internal/utils/logging.py

```diff
@@ -21,11 +21,26 @@
 
 
 try:
-    from pip._vendor import colorama
+    # Use "import as" and set colorama in the else clause to avoid mypy
+    # errors and get the following correct revealed type for colorama:
+    # `Union[_importlib_modulespec.ModuleType, None]`
+    # Otherwise, we get an error like the following in the except block:
+    #  > Incompatible types in assignment (expression has type "None",
+    #   variable has type Module)
+    # TODO: eliminate the need to use "import as".
+    from pip._vendor import colorama as _colorama
 # Lots of different errors can come from this, including SystemError and
 # ImportError.
 except Exception:
     colorama = None
+else:
+    # Import Fore explicitly rather than accessing below as colorama.Fore
+    # to avoid the following error running mypy:
+    # > Module has no attribute "Fore"
+    # TODO: eliminate the need to import Fore.
```

Comment:
This looks like python/mypy#3500. Specifically, it looks like when the `Union` is narrowed to `ModuleType` on line 168 below it has already lost information about the attributes available for the module (colorama). This is reinforced by adding

```
reveal_type(_colorama)
reveal_type(_colorama.Fore)
```

above this line, which outputs

```
src/pip/_internal/utils/logging.py:37: error: Revealed type is '_importlib_modulespec.ModuleType'
src/pip/_internal/utils/logging.py:38: error: Revealed type is 'pip._vendor.colorama.ansi.AnsiFore'
```

during the mypy check as expected. Can we add a link to the issue?

Reply:
I thought there was a better issue for this. As above, I suspect this whole area is being / needs reworking, so it's not one isolated issue that needs to be fixed. But I can pick something out or reference an umbrella issue.